File: src/pumila/window.cc
#include <pumila/window.h>
#include <pumila/model_base.h>
#include "drawing.h"
#include <iostream>
#include <sstream>

#ifdef PUMILA_SDL2
#include <SDL.h>
#endif

namespace PUMILA_NS {
Window::Window(const std::vector<std::shared_ptr<GameSim>> &sim)
    : sim(sim), key_state(this) {
#ifdef PUMILA_SDL2
    SDL_Window *sdl_window = nullptr;
    SDL_Renderer *sdl_renderer = nullptr;
    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        std::cerr << "Couldn't initialize SDL: " << SDL_GetError() << std::endl;
    } else {
        sdl_window =
            SDL_CreateWindow("Pumila", SDL_WINDOWPOS_UNDEFINED,
                             SDL_WINDOWPOS_UNDEFINED, WIDTH, HEIGHT, 0);
        sdl_window_p = static_cast<void *>(sdl_window);
        if (!sdl_window) {
            std::cerr << "Couldn't create window: " << SDL_GetError()
                      << std::endl;
        } else {
            SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "linear");
            sdl_renderer =
                SDL_CreateRenderer(sdl_window, -1, SDL_RENDERER_ACCELERATED);
            sdl_renderer_p = static_cast<void *>(sdl_renderer);
            if (!sdl_renderer) {
                std::cerr << "Couldn't create renderer: " << SDL_GetError()
                          << std::endl;
            }
        }
    }
    TTF_Init();
    TTF_Font *ttf_font = TTF_OpenFont("Roboto-Regular.ttf", 24);
    ttf_font_p = static_cast<void *>(ttf_font);
    TTF_Font *ttf_font_sm = TTF_OpenFont("Roboto-Regular.ttf", 16);
    ttf_font_sm_p = static_cast<void *>(ttf_font_sm);
    if (!ttf_font) {
        std::cerr << "Font not found" << std::endl;
    }
#else
    std::cerr << "Pumila not built with SDL2" << std::endl;
#endif
}

Window::~Window() { quit(); }
void Window::quit() {
#ifdef PUMILA_SDL2
    if (isRunning()) {
        SDL_Quit();
        sdl_window_p = nullptr;
        sdl_renderer_p = nullptr;
    }
#endif
}
bool Window::isRunning() const { return sdl_window_p && sdl_renderer_p; }

void Window::step(bool sim_step) {
#ifdef PUMILA_SDL2
    if (isRunning()) {
        handleEvent();
    }
    key_state.keyFrame();
    if (sim_step && state == WindowState::game) {
        for (const auto &s : sim) {
            s->step();
        }
    }
    if (isRunning()) {
        draw();
        SDL_Delay(16);
    }
    phase_t++;
    switch (state) {
    case WindowState::ready:
        if (phase_t >= READY_T) {
            state = WindowState::game;
            phase_t = 0;
        }
        break;
    case WindowState::game:
        for (std::size_t i = 0; i < sim.size(); i++) {
            std::shared_lock lock(sim[i]->field_m);
            if (sim[i]->is_over) {
                state = WindowState::finish;
                phase_t = 0;
            }
        }
        break;
    default:
        break;
    }
#endif
}

void Window::draw() {
#ifdef PUMILA_SDL2
    using namespace PUMILA_NS::drawing;
    SDL_Renderer *sdl_renderer = static_cast<SDL_Renderer *>(sdl_renderer_p);
    TTF_Font *ttf_font = static_cast<TTF_Font *>(ttf_font_p);
    TTF_Font *ttf_font_sm = static_cast<TTF_Font *>(ttf_font_sm_p);
    std::ostringstream ss;
    int text_w, text_h;
    SDL_Texture *text;
    SDL_Rect rect;

    SDL_SetRenderDrawBlendMode(sdl_renderer, SDL_BLENDMODE_BLEND);

    // 背景
    SDL_SetRenderDrawColor(sdl_renderer, 255, 255, 255, 255);
    SDL_RenderClear(sdl_renderer);

    for (std::size_t i = 0; i < sim.size(); i++) {
        std::shared_lock lock_f(sim[i]->field_m);
        std::lock_guard lock_s(sim[i]->step_m);
        // 枠
        SDL_SetRenderDrawColor(sdl_renderer, 0, 0, 0, 255);
        SDL_Rect field_rect = {FIELD_X[i], FIELD_Y - PUYO_SIZE * 12,
                               PUYO_SIZE * 6, PUYO_SIZE * 12};
        SDL_RenderDrawRect(sdl_renderer, &field_rect);

        FieldState2 *field = &*sim[i]->field;
        auto fall_phase =
            dynamic_cast<GameSim::FallPhase *>(sim[i]->phase.get());
        if (fall_phase) {
            field = &fall_phase->display_field;
        }
        if (sim[i]->isFreePhase()) {
            auto current_pair = field->next().get();
            drawPuyo(current_pair.bottom, current_pair.bottomX(),
                     field->getNextHeight(current_pair).first, i, false);
            drawPuyo(current_pair.top, current_pair.topX(),
                     field->getNextHeight(current_pair).second, i, false);
            drawPuyo(current_pair.bottom, current_pair.bottomX(),
                     current_pair.bottomY(), i, true);
            drawPuyo(current_pair.top, current_pair.topX(), current_pair.topY(),
                     i, true);
        }
        std::size_t next_p = sim[i]->isFreePhase() ? 1 : 0;
        drawPuyo(field->next()[next_p].bottom, 6.5, 10.5, i, true);
        drawPuyo(field->next()[next_p].top, 6.5, 11.5, i, true);
        drawPuyo(field->next()[next_p + 1].bottom, 8, 9.5, i, true);
        drawPuyo(field->next()[next_p + 1].top, 8, 10.5, i, true);
        for (std::size_t y = 0; y < FieldState2::HEIGHT; y++) {
            for (std::size_t x = 0; x < FieldState2::WIDTH; x++) {
                drawPuyo(field->field().get(x, y), x, y, i, true);
            }
        }

        if (ttf_font) {
            text = drawText(sdl_renderer,
                            sim[i]->name.empty() ? "player" : sim[i]->name,
                            ttf_font, {0, 0, 0, 255}, &text_w, &text_h);
            rect = {FIELD_X[i] + PUYO_SIZE * 3 - text_w / 2, FIELD_Y + 5,
                    text_w, text_h};
            SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
            SDL_DestroyTexture(text);

            ss << field->totalScore();
            text = drawText(sdl_renderer, ss.str(), ttf_font, {0, 0, 0, 255},
                            &text_w, &text_h);
            rect = {FIELD_X[i] + PUYO_SIZE * 6 - text_w - 10, FIELD_Y + 35,
                    text_w, text_h};
            SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
            SDL_DestroyTexture(text);
            ss.str("");


            int ready = field->garbage().getReady();
            bool red = ready >= 30;
            auto garbage_gauge = [](int garbage) {
                return static_cast<int>(
                    PUYO_SIZE * 3 *
                    std::log1p(static_cast<double>(garbage > 0 ? garbage : 0)) /
                    std::log(30.0));
            };
            if (ready > 0) {
                drawPuyo(Puyo::garbage, 0.5, 13.5, i, true);
                ss << "x " << ready;
                text = drawText(sdl_renderer, ss.str(), ttf_font,
                                red ? SDL_Color{255, 0, 0, 255}
                                    : SDL_Color{0, 0, 0, 255},
                                &text_w, &text_h);
                rect = {FIELD_X[i] + static_cast<int>(PUYO_SIZE * 1.5) + 5,
                        FIELD_Y - PUYO_SIZE * 14 - 15, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");

                SDL_SetRenderDrawColor(sdl_renderer, (red ? 255 : 0), 0, 0,
                                       255);
                rect = {FIELD_X[i], FIELD_Y - PUYO_SIZE * 13 - 10,
                        garbage_gauge(ready), 3};
                SDL_RenderFillRect(sdl_renderer, &rect);
            }

            int current = field->garbage().getCurrent();
            if (current > 0) {
                ss << "- " << current;
                text = drawText(sdl_renderer, ss.str(), ttf_font,
                                SDL_Color{0, 190, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + PUYO_SIZE * 6 - text_w - 10,
                        FIELD_Y - PUYO_SIZE * 14 - 15, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");

                SDL_SetRenderDrawColor(sdl_renderer, 0, 190, 0, 255);
                if (ready > 0) {
                    rect = {FIELD_X[i] + garbage_gauge(ready - current),
                            FIELD_Y - PUYO_SIZE * 13 - 10, garbage_gauge(ready),
                            3};
                    SDL_RenderFillRect(sdl_renderer, &rect);
                }
                if (current - ready > 0) {
                    rect = {FIELD_X[i] + PUYO_SIZE * 6 -
                                garbage_gauge(current - ready),
                            FIELD_Y - PUYO_SIZE * 13 - 10,
                            garbage_gauge(current - ready), 3};
                    SDL_RenderFillRect(sdl_renderer, &rect);
                }
            }

            if (sim[i]->model) {
                ss << "ActionCoeff";
                text = drawText(sdl_renderer, ss.str(), ttf_font_sm,
                                {0, 0, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + PUYO_SIZE * 6 + 10,
                        FIELD_Y - PUYO_SIZE * 1 - 20, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");

                ss << sim[i]->model->actionCoeff();
                text = drawText(sdl_renderer, ss.str(), ttf_font_sm,
                                {0, 0, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + PUYO_SIZE * 9 - text_w,
                        FIELD_Y - PUYO_SIZE * 1, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");
            }
            if (fall_phase && fall_phase->current_chain) {
                ss << fall_phase->current_chain->scoreA() << " x "
                   << fall_phase->current_chain->scoreB();
                text = drawText(sdl_renderer, ss.str(), ttf_font,
                                {0, 0, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + PUYO_SIZE * 6 - text_w - 10, FIELD_Y + 60,
                        text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");

                ss << fall_phase->current_chain->chain_num;
                text = drawText(sdl_renderer, ss.str(), ttf_font,
                                {0, 0, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + static_cast<int>(PUYO_SIZE * 7.5) -
                            text_w / 2,
                        FIELD_Y - PUYO_SIZE * 3 - 30, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");

                ss << "chain";
                if (fall_phase->current_chain->chain_num >= 2) {
                    ss << "s";
                }
                ss << "!";
                text = drawText(sdl_renderer, ss.str(), ttf_font_sm,
                                {0, 0, 0, 255}, &text_w, &text_h);
                rect = {FIELD_X[i] + static_cast<int>(PUYO_SIZE * 7.5) -
                            text_w / 2,
                        FIELD_Y - PUYO_SIZE * 3, text_w, text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
                ss.str("");
            }
        }
    }
    if (ttf_font) {

        switch (state) {
        case WindowState::ready:
            text = drawText(sdl_renderer, "Ready?", ttf_font, {0, 0, 0, 255},
                            &text_w, &text_h);
            rect = {WIDTH / 2 - text_w / 2, HEIGHT / 2 - text_h / 2, text_w,
                    text_h};
            SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
            SDL_DestroyTexture(text);
            break;
        case WindowState::game:
            if (phase_t < GO_T) {
                text = drawText(sdl_renderer, "Go!", ttf_font, {0, 0, 0, 255},
                                &text_w, &text_h);
                rect = {WIDTH / 2 - text_w / 2, HEIGHT / 2 - text_h / 2, text_w,
                        text_h};
                SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
                SDL_DestroyTexture(text);
            }
            break;
        case WindowState::finish:
            text = drawText(sdl_renderer, "Finish", ttf_font, {0, 0, 0, 255},
                            &text_w, &text_h);
            rect = {WIDTH / 2 - text_w / 2, HEIGHT / 2 - text_h / 2, text_w,
                    text_h};
            SDL_RenderCopy(sdl_renderer, text, NULL, &rect);
            SDL_DestroyTexture(text);
            break;
        default:
            std::cerr << "invalid state " << static_cast<int>(state)
                      << std::endl;
            state = WindowState::ready;
            break;
        }
    }

    SDL_RenderPresent(sdl_renderer);
#endif
}
void Window::drawPuyo(Puyo p, double x, double y, int i, bool not_ghost) {
#ifdef PUMILA_SDL2
    if (isRunning() && p != Puyo::none) {
        using namespace PUMILA_NS::drawing;
        SDL_Renderer *sdl_renderer =
            static_cast<SDL_Renderer *>(sdl_renderer_p);
        Points points;
        SDL_SetRenderDrawColor(sdl_renderer, col(p).r, col(p).g, col(p).b,
                               not_ghost ? 255 : 100);
        points = drawCircle(puyoX(x, i), puyoY(y),
                            PUYO_SIZE * (not_ghost ? 0.9 : 0.5) / 2);
        SDL_RenderDrawPoints(sdl_renderer, points.data(), points.size());
    }
#endif
}

void Window::handleEvent() {
#ifdef PUMILA_SDL2
    SDL_Event event;
    while (SDL_PollEvent(&event)) {
        switch (event.type) {
        case SDL_QUIT:
            SDL_Quit();
            sdl_window_p = nullptr;
            sdl_renderer_p = nullptr;
            return;
        case SDL_KEYDOWN:
        case SDL_KEYUP: {
            if (!event.key.repeat) {
                bool is_down = event.type == SDL_KEYDOWN;
                key_state.key_repeat_wait = 0;
                switch (event.key.keysym.scancode) {
                case SDL_SCANCODE_UP:
                case SDL_SCANCODE_W:
                    key_state.quick_drop = is_down;
                    break;
                case SDL_SCANCODE_DOWN:
                case SDL_SCANCODE_S:
                    key_state.soft_drop = is_down;
                    break;
                case SDL_SCANCODE_LEFT:
                case SDL_SCANCODE_A:
                    key_state.left = is_down;
                    break;
                case SDL_SCANCODE_RIGHT:
                case SDL_SCANCODE_D:
                    key_state.right = is_down;
                    break;
                case SDL_SCANCODE_KP_DIVIDE:
                case SDL_SCANCODE_N:
                    key_state.rot_left = is_down;
                    break;
                case SDL_SCANCODE_KP_MULTIPLY:
                case SDL_SCANCODE_M:
                    key_state.rot_right = is_down;
                    break;
                case SDL_SCANCODE_P:
                    key_state.reset = is_down;
                    break;
                default:
                    break;
                }
            }
            break;
        }
        default:
            break;
        }
    }
#endif
}
void Window::KeyState::keyFrame() {
    if (reset) {
        for (const auto &s : win->sim) {
            s->reset();
        }
        win->state = WindowState::ready;
        win->phase_t = 0;
    }
    reset = false;
    if (quick_drop && win->state == WindowState::game) {
        for (const auto &s : win->sim) {
            if (!s->hasModel()) {
                s->quickDrop();
            }
        }
    }
    quick_drop = false;
    if (soft_drop && win->state == WindowState::game) {
        for (const auto &s : win->sim) {
            if (!s->hasModel()) {
                s->softDrop();
            }
        }
    }
    if (left && win->state == WindowState::game) {
        if (!key_repeat_wait) {
            for (const auto &s : win->sim) {
                if (!s->hasModel()) {
                    s->movePair(-1);
                }
            }
        }
    }
    if (right && win->state == WindowState::game) {
        if (!key_repeat_wait) {
            for (const auto &s : win->sim) {
                if (!s->hasModel()) {
                    s->movePair(1);
                }
            }
        }
    }
    if (rot_left && win->state == WindowState::game) {
        for (const auto &s : win->sim) {
            if (!s->hasModel()) {
                s->rotPair(-1);
            }
        }
        rot_left = false;
    }
    if (rot_right && win->state == WindowState::game) {
        for (const auto &s : win->sim) {
            if (!s->hasModel()) {
                s->rotPair(1);
            }
        }
        rot_right = false;
    }
    if (key_repeat_wait == 0) {
        key_repeat_wait = KEY_REPEAT;
    } else {
        key_repeat_wait--;
    }
}

} // namespace PUMILA_NS

File: pumila/window.py
import pygame
from pygame.locals import *
from enum import Enum
import pypumila
from typing import List, Tuple
import math

WIDTH = 700
HEIGHT = 560
PUYO_SIZE = 30
FIELD_X = [50, 400]
FIELD_Y = HEIGHT - 100


def puyo_x(x: float, i: int) -> int:
    return round(FIELD_X[i] + PUYO_SIZE * (2 * x + 1) / 2)


def puyo_y(y: float) -> int:
    return round(FIELD_Y - PUYO_SIZE * (2 * y + 1) / 2)


class KeyState:
    soft_drop: bool
    quick_drop: bool
    left: bool
    right: bool
    rot_left: bool
    rot_right: bool
    reset: bool
    key_repeat_wait: int

    def __init__(self) -> None:
        self.soft_drop = False
        self.quick_drop = False
        self.left = False
        self.right = False
        self.rot_left = False
        self.rot_right = False
        self.reset = False
        self.key_repeat_wait = 0


class WindowPhase(Enum):
    READY = 0
    GAME = 1
    FINISH = 2


class Window:
    window_init: bool = False
    screen: pygame.Surface
    font: pygame.font.Font
    key_state: KeyState
    phase_t: int
    phase: WindowPhase
    sim: List[pypumila.GameSim]
    is_player: List[bool]

    def __init__(self) -> None:
        if Window.window_init:
            raise RuntimeError("pygame already initialized")
        pygame.init()
        pygame.key.set_repeat()
        self.screen = pygame.display.set_mode((700, 560))
        pygame.font.init()
        self.font = pygame.font.Font("Roboto-Regular.ttf", 24)
        self.font_sm = pygame.font.Font("Roboto-Regular.ttf", 16)
        Window.window_init = True
        self.key_state = KeyState()
        self.phase_t = 0
        self.phase = WindowPhase.READY
        self.sim = []
        self.is_player = []

    def set_sim(self, sim: List[Tuple[pypumila.GameSim, bool]]) -> None:
        assert len(sim) <= 2
        self.sim = [s[0] for s in sim]
        self.is_player = [s[1] for s in sim]

    def step(self) -> None:
        self.handle_event()
        self.key_frame()
        self.draw()

    def handle_event(self) -> None:
        for event in pygame.event.get():
            if event.type in [KEYDOWN, KEYUP]:
                self.key_state.key_repeat_wait = 0
                is_down = event.type == KEYDOWN
                if event.key in [K_w, K_UP]:
                    self.key_state.quick_drop = is_down
                elif event.key in [K_s, K_DOWN]:
                    self.key_state.soft_drop = is_down
                elif event.key in [K_a, K_LEFT]:
                    self.key_state.left = is_down
                elif event.key in [K_d, K_RIGHT]:
                    self.key_state.right = is_down
                elif event.key in [K_n, K_KP_DIVIDE]:
                    self.key_state.rot_left = is_down
                elif event.key in [K_m, K_KP_MULTIPLY]:
                    self.key_state.rot_right = is_down
                elif event.key in [K_p]:
                    self.key_state.reset = is_down

    def key_frame(self) -> None:
        if self.key_state.reset:
            for s in self.sim:
                s.reset()
            self.phase = WindowPhase.READY
            self.phase_t = 0
        self.key_state.reset = False

        if self.key_state.quick_drop and self.phase == WindowPhase.GAME:
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.quick_drop()
        self.key_state.quick_drop = False

        if self.key_state.soft_drop and self.phase == WindowPhase.GAME:
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.soft_drop()
        self.key_state.soft_drop = False

        if (
            self.key_state.left
            and self.phase == WindowPhase.GAME
            and not self.key_state.key_repeat_wait
        ):
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.move_pair(-1)

        if (
            self.key_state.right
            and self.phase == WindowPhase.GAME
            and not self.key_state.key_repeat_wait
        ):
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.move_pair(1)

        if self.key_state.rot_left and self.phase == WindowPhase.GAME:
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.rot_pair(-1)
        self.key_state.rot_left = False

        if self.key_state.rot_right and self.phase == WindowPhase.GAME:
            for s, p in zip(self.sim, self.is_player):
                if p:
                    s.rot_pair(1)
        self.key_state.rot_right = False

        if self.key_state.key_repeat_wait == 0:
            self.key_state.key_repeat_wait = 10
        else:
            self.key_state.key_repeat_wait -= 1

    def draw(self) -> None:
        self.screen.fill(pygame.Color(255, 255, 255))

        black = pygame.Color(0, 0, 0)
        for i, s in enumerate(self.sim):
            pygame.draw.rect(
                self.screen,
                black,
                pygame.Rect(
                    FIELD_X[i], FIELD_Y - PUYO_SIZE * 12, PUYO_SIZE * 6, PUYO_SIZE * 12
                ),
                width=1,
            )

            field = s.fall_phase_display_field()
            if field is None:
                field = s.field_copy()
            next_p = 0
            if s.phase_get() == pypumila.GameSim.PhaseEnum.free:
                next_p = 1
                pair = field.get_next(0)
                self.draw_puyo(
                    pair.bottom,
                    pair.bottom_x(),
                    field.get_next_height(pair)[0],
                    i,
                    False,
                )
                self.draw_puyo(
                    pair.top, pair.top_x(), field.get_next_height(pair)[1], i, False
                )
                self.draw_puyo(pair.bottom, pair.bottom_x(), pair.bottom_y(), i, True)
                self.draw_puyo(pair.top, pair.top_x(), pair.top_y(), i, True)
            next1 = field.get_next(next_p)
            self.draw_puyo(next1.bottom, 6.5, 10.5, i, True)
            self.draw_puyo(next1.top, 6.5, 11.5, i, True)
            next2 = field.get_next(next_p + 1)
            self.draw_puyo(next2.bottom, 8, 9.5, i, True)
            self.draw_puyo(next2.top, 8, 10.5, i, True)

            for y in range(13):
                for x in range(6):
                    self.draw_puyo(field.get(x, y), x, y, i, True)

            score_text = self.font.render(str(field.total_score()), True, black)
            self.screen.blit(
                score_text,
                (
                    FIELD_X[i] + PUYO_SIZE * 6 - score_text.get_width() - 10,
                    FIELD_Y + 35,
                ),
            )

            def garbage_gauge(garbage: int) -> int:
                if garbage > 0:
                    return round(PUYO_SIZE * 3 * math.log(1 + garbage) / math.log(30))
                return 0

            ready = field.get_garbage_num_total()
            red = ready >= 30
            if ready > 0:
                self.draw_puyo(pypumila.Puyo.garbage, 0.5, 13.5, i, True)
                gb_color = black
                if red:
                    gb_color = pygame.Color(255, 0, 0)
                gb_text = self.font.render(f"x {ready}", True, gb_color)
                self.screen.blit(
                    gb_text,
                    (
                        FIELD_X[i] + round(PUYO_SIZE * 1.5) + 5,
                        FIELD_Y - PUYO_SIZE * 14 - 15,
                    ),
                )
                pygame.draw.rect(
                    self.screen,
                    gb_color,
                    pygame.Rect(
                        FIELD_X[i],
                        FIELD_Y - PUYO_SIZE * 13 - 10,
                        garbage_gauge(ready),
                        3,
                    ),
                )

            if (
                s.phase_get() == pypumila.GameSim.PhaseEnum.fall
                and s.fall_phase_current_chain() < len(s.current_step().chains)
            ):
                current_chain = s.current_step().chains[s.fall_phase_current_chain()]
                current_sc_text = self.font.render(
                    f"{current_chain.score_a()} x {current_chain.score_b()}",
                    True,
                    black,
                )
                self.screen.blit(
                    current_sc_text,
                    (
                        FIELD_X[i] + PUYO_SIZE * 6 - current_sc_text.get_width() - 10,
                        FIELD_Y + 60,
                    ),
                )

                current_chain_text = self.font.render(
                    str(current_chain.chain_num),
                    True,
                    black,
                )
                self.screen.blit(
                    current_chain_text,
                    (
                        round(
                            FIELD_X[i]
                            + PUYO_SIZE * 7.5
                            - current_chain_text.get_width() / 2
                        ),
                        FIELD_Y - PUYO_SIZE * 3 - 30,
                    ),
                )

                chain_text = self.font_sm.render(
                    "chains!" if current_chain.chain_num >= 2 else "chain!",
                    True,
                    black,
                )
                self.screen.blit(
                    chain_text,
                    (
                        round(
                            FIELD_X[i] + PUYO_SIZE * 7.5 - chain_text.get_width() / 2
                        ),
                        FIELD_Y - PUYO_SIZE * 3,
                    ),
                )

        if self.phase == WindowPhase.READY:
            text = self.font.render("Ready?", True, black)
            if self.phase_t >= 60:
                self.phase_t = 0
                self.phase = WindowPhase.GAME
        elif self.phase == WindowPhase.GAME:
            if self.phase_t < 60:
                text = self.font.render("Go!", True, black)
            else:
                text = None
            for s in self.sim:
                if s.is_over:
                    self.phase_t = 0
                    self.phase = WindowPhase.FINISH
        elif self.phase == WindowPhase.FINISH:
            text = self.font.render("Finish", True, black)
        if text is not None:
            self.screen.blit(
                text,
                (
                    round(WIDTH / 2 - text.get_width() / 2),
                    round(HEIGHT / 2 - text.get_height() / 2),
                ),
            )

        pygame.display.flip()

        self.phase_t += 1

    def draw_puyo(self, p: pypumila.Puyo, x: float, y: float, i: int, not_ghost: bool):
        if p == pypumila.Puyo.none:
            return
        elif p == pypumila.Puyo.red:
            if not_ghost:
                color = pygame.Color(255, 120, 120)
            else:
                color = pygame.Color(255, 200, 200)
        elif p == pypumila.Puyo.blue:
            if not_ghost:
                color = pygame.Color(120, 120, 255)
            else:
                color = pygame.Color(200, 200, 255)
        elif p == pypumila.Puyo.green:
            if not_ghost:
                color = pygame.Color(120, 255, 120)
            else:
                color = pygame.Color(200, 255, 200)
        elif p == pypumila.Puyo.yellow:
            if not_ghost:
                color = pygame.Color(220, 220, 0)
            else:
                color = pygame.Color(255, 255, 120)
        elif p == pypumila.Puyo.purple:
            if not_ghost:
                color = pygame.Color(255, 0, 255)
            else:
                color = pygame.Color(255, 120, 255)
        elif p == pypumila.Puyo.garbage:
            color = pygame.Color(190, 190, 190)

        if not_ghost:
            radius = PUYO_SIZE * 0.9 / 2
        else:
            radius = PUYO_SIZE * 0.5 / 2

        pygame.draw.circle(self.screen, color, (puyo_x(x, i), puyo_y(y)), round(radius))
